### PR TITLE
fix: sessions cannot be edited after cfs ends

### LINKED ISSUE #6726: Session & Speaker: User can change his session/speaker details after CFS is closed
**Describe the bug**

User can change his session/speaker details after CFS is closed.

**To Reproduce**

Steps to reproduce the behaviour:

1. Submit a Proposal and Confirm it.
2. Wait for CFS to get closed and then edit your proposal.
3. See error

**Expected behaviour**

User should not have the right to edit the proposal once CFS is ended.

See Also : https://github.com/fossasia/open-event-frontend/issues/3768#issuecomment-571141434

### PR DESCRIPTION
**Pull request by codedsun**
Fixes #6726 

#### Short description of what this resolves:
fix: sessions cannot be edited after cfs ends

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

### CODE REVIEW

**Comment by codedsun** [comment]
@mrsaicharan1 - Yes, I am fixing the dredd test

File: app/api/sessions.py [hunk 1]
```diff
@@ -148,6 +150,11 @@ def before_update_object(self, session, data, view_kwargs):
         if session.is_locked and data.get('is_locked') == session.is_locked:
             raise ForbiddenException({'source': '/data/attributes/is-locked'}, "Locked sessions cannot be edited")
 
+        speakers_call = safe_query(self, SpeakersCall, 'event_id', session.event_id, 'event-id')
+        if speakers_call.ends_at.replace(tzinfo=None) < datetime.now().replace(tzinfo=None):
```
Comment (mrsaicharan1) [inline app/api/sessions.py]: No need to replace the `tzinfo`. Timezone does not play any role here.

Comment (mrsaicharan1) [inline app/api/sessions.py]: if speakers_call.ends_at.replace(tzinfo=None) <= datetime.now().replace(tzinfo=None):

Comment (codedsun) [reply app/api/sessions.py]: ?
The speakers calls `ends at` is timezone aware
`  ends_at = db.Column(db.DateTime(timezone=True), nullable=False)`

Comment (mrsaicharan1) [reply app/api/sessions.py]: Suppose the event was created in the GMT timezone and you're in a CST timezone, would the speakers call time be independent of the timezone or would it be dependent when you're trying to access it?

Comment (codedsun) [reply app/api/sessions.py]: It should be dependent on my timezone when I am accessing it.

File: app/api/sessions.py [hunk 2]
```diff
@@ -148,6 +150,11 @@ def before_update_object(self, session, data, view_kwargs):
         if session.is_locked and data.get('is_locked') == session.is_locked:
             raise ForbiddenException({'source': '/data/attributes/is-locked'}, "Locked sessions cannot be edited")
 
+        speakers_call = safe_query(self, SpeakersCall, 'event_id', session.event_id, 'event-id')
+        if speakers_call.ends_at.replace(tzinfo=None) <= datetime.now().replace(tzinfo=None):
```
Comment (iamareebjamal) [inline app/api/sessions.py]: Now I won't be able to delete the session as well.

Second, you have made it timezone unaware

File: app/api/sessions.py (hunk 1 above)
Comment (codedsun) [reply app/api/sessions.py]: It should be dependent. 

Like for GMT person - it ending at 5:00AM
Like for an CST person - it's ending at 3:00AM 

So for the CST person, editing is allowed < 3:00AM and for GMT people < 5:00 AM

File: app/api/sessions.py (hunk 2 above)
Comment (codedsun) [reply app/api/sessions.py]: @iamareebjamal - Shall the sessions be deleted after the CFS is ended?

File: app/api/sessions.py (hunk 1 above)
Comment (iamareebjamal) [reply app/api/sessions.py]: Yes, so why are you removing timezone from date?

File: app/api/sessions.py (hunk 2 above)
Comment (iamareebjamal) [reply app/api/sessions.py]: Admin/Co-organizer should be able to edit and delete both

File: app/api/sessions.py (hunk 1 above)
Comment (codedsun) [reply app/api/sessions.py]: Yes, yes undoing it.

**Comment by codedsun** [comment]
@iamareebjamal - Review

**Comment by kushthedude** [comment]
> @iamareebjamal - Review

Read the complete issue, Speaker details needs to be locked as well. Or if you are partially fixing than don't mention the `fixes` keyword!

**Comment by codedsun** [comment]
> Read the complete issue, Speaker details needs to be locked as well

I have read the issue, the speaker details can't be locked currently as it's related to the user not the CFS of the event. Once the speaker is created the same profile is applied for all the cfs of every event. 
@kushthedude

**Comment by kushthedude** [comment]
Remove the keyword fixes then.

**Comment by codedsun** [comment]
@kushthedude - What I looked upon is for every event a new speaker profile is generated.

So from the frontend side, you have to change the request body and pass the event id for that

**Comment by codedsun** [comment]
@iamareebjamal @kushthedude @mrsaicharan1 - Please review. Thanks

**Comment by codedsun** [comment]
@iamareebjamal - Review please thanks.

File: app/api/speakers.py [hunk 3]
```diff
@@ -132,6 +135,14 @@ def before_update_object(self, speaker, data, view_kwargs):
         :param view_kwargs:
         :return:
         """
+        speakers_call = SpeakersCall.query.filter_by(event_id=speaker.event_id).one()
+        if speakers_call and not (has_access('is_admin') or has_access('is_organizer', event_id=speaker.event_id) or
+                                  has_access('is_coorganizer', event_id=speaker.event_id)):
+            speakers_call_tz = speakers_call.ends_at.tzinfo
+            if speakers_call.ends_at <= datetime.now().replace(tzinfo=speakers_call_tz):
```
Comment (iamareebjamal) [inline app/api/speakers.py]: Duplicate logic

**Comment by codedsun** [comment]
@iamareebjamal - Review now

File: app/api/helpers/speaker.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+def can_edit_after_cfs_ends(event_id):
+    """
+    Method to check that user has permission to edit the speaker or session
+    after the CFS ends
+    """
+    speakers_call = SpeakersCall.query.filter_by(event_id=event_id).one()
+    if speakers_call:
+        speakers_call_tz = speakers_call.ends_at.tzinfo
+        if speakers_call.ends_at <= datetime.now().replace(tzinfo=speakers_call_tz) \
```
Comment (iamareebjamal) [inline app/api/helpers/speaker.py]: Just use datetime.utcnow()

Comment (iamareebjamal) [inline app/api/helpers/speaker.py]: Just return not of this expression, no need for explicit returns of boolean

Comment (codedsun) [reply app/api/helpers/speaker.py]: @iamareebjamal - Won't work

`TypeError: can't compare offset-naive and offset-aware datetimes`

Comment (codedsun) [reply app/api/helpers/speaker.py]: We have to compare the current time (with the timezone of the speaker call) and the speaker call end date which already is timezone aware

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/api/helpers/speaker.py  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4845029&bid=16082287)

**Comment by codedsun** [comment]
@iamareebjamal - Ping!

File: app/api/helpers/speaker.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+from datetime import datetime
+
+
+def can_edit_after_cfs_ends(event_id):
+    """
+    Method to check that user has permission to edit the speaker or session
+    after the CFS ends
+    """
+    speakers_call = SpeakersCall.query.filter_by(event_id=event_id).one()
```
Comment (iamareebjamal) [inline app/api/helpers/speaker.py]: Will not filter out deleted speaker calls

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/api/helpers/speaker.py  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4845029&bid=16082287)

**Comment by codedsun** [comment]
@iamareebjamal - Review

**Comment by codedsun** [comment]
@iamareebjamal - Thankyou lord 🥇😆

**Pull request merged**